_posts/2024-08-07-accessing-kafka-with-gateway-api.md

```diff
@@ -0,0 +1,627 @@
+
+![Architecture with TLSRoutes](/assets/images/posts/2024-08-07-tls-routes.png)
+
+## Putting It Into Practice
+
+The rest of this blog post will walk through how to use `TLSRoute`s to access a Strimzi-managed Kafka cluster from outside of your Kubernetes cluster. We will use a [KIND](https://kind.sigs.k8s.io/) cluster, which allows us to run a Kubernetes cluster in docker containers on our local machine, and we will use Envoy Gateway as our implementation of the Gateway API.
+
+### KIND Cluster Setup
+
+First, let's create the KIND cluster using the following `kind-config.yaml` file:
+
+```
+kind: Cluster
+apiVersion: kind.x-k8s.io/v1alpha4
+nodes:
+- kubeadmConfigPatches:
+  - |
+    kind: InitConfiguration
+    nodeRegistration:
+      kubeletExtraArgs:
+        node-labels: "ingress-ready=true"
+  extraPortMappings:
+  - containerPort: 30992
+    hostPort: 9092
+    protocol: TCP
+```
+
+If you inspect the `kind-config.yaml` file, you will notice the port mapping of `hostPort: 9092` being mapped to `containerPort: 30992`. That means that the port 9092 on your own laptop will be forwarded by docker to port 30992 on the Kubernetes Node (which, in KIND, is just a docker container running on your laptop). We will use this fact when installing Envoy Gateway.
+
+```
+kind create cluster --name strimzi-gw-api --config kind-config.yaml
+```
+
+Next, let's set up Envoy Gateway. First, we will use `helm` to install it.
+
+```
+helm upgrade --install envoygateway oci://docker.io/envoyproxy/gateway-helm \
+    --version v1.0.1 \
+    --namespace envoy-gateway-system \
+    --create-namespace
+```
+
+Once the installation process is complete, we need to deploy a `Gateway` with a specific `GatewayClass` that will listen on the correct `NodePort`s. We'll refer to this `Gateway` later when we create `TLSRoute`s to access our Kafka cluster. Let's apply the following file:
+
+```
+apiVersion: gateway.networking.k8s.io/v1
+kind: GatewayClass
+metadata:
+  name: my-gateway-class
+  namespace: envoy-gateway-system
+spec:
+  controllerName: gateway.envoyproxy.io/gatewayclass-controller
+  parametersRef:
+    group: gateway.envoyproxy.io
+    kind: EnvoyProxy
+    name: my-proxy-config
+    namespace: envoy-gateway-system
+---
+apiVersion: gateway.envoyproxy.io/v1alpha1
+kind: EnvoyProxy
+metadata:
+  name: my-proxy-config
+  namespace: envoy-gateway-system
+spec:
+  provider:
+    type: Kubernetes
+    kubernetes:
+      envoyDeployment:
+        replicas: 1
+      envoyService:
+        type: NodePort
+        patch:
+          value:
+            spec:
+              ports:
+              # Port 9092 on your laptop gets forwarded to NodePort 30992 on the KIND cluster.
+              - name: kafka-port
+                nodePort: 30992
+                port: 9092
+                protocol: TCP
+                targetPort: 9092
+---
+apiVersion: gateway.networking.k8s.io/v1
+kind: Gateway
+metadata:
+  name: my-gateway
+  namespace: default
```

Comment:
In many Kubernetes clusters, the `default` namespace has special properties and might not work out of the box. Would it be better to create some new namespace and use it instead of `default`?